python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -740,3 +742,83 @@ def with_reconstruction_metadata(self, state_metadata: Mapping[str, Any]) -> Sel
                 **state_metadata,
             },
         )
+
+
+def map_asset_specs(defs: Definitions, fn: Callable[[AssetSpec], AssetSpec]) -> "Definitions":
+    """Applies the given mapping function to every asset spec within this Definitions object,
+    including both external asset specs and asset specs within AssetsDefinitions.
+
+    Returns a new Definitions object with the transformed assets; does not mutate this object.
+
+    If the mapping function produces asset specs with new asset keys, downstream assets and
+    asset checks will be updated so that they depend on the new key.
+
+    Does not support Definitions  objects that contain CacheableAssetsDefinitions.
+
+    Args:
+        fn (Callable[[AssetSpec], AssetSpec]): A function that accepts an AssetSpec and returns
+            a new AssetSpec.
+
+    Returns:
+        Definitions: A new Definitions object with the transformed assets.
+    """
+    # Don't want to unintentionally trigger a fetch from an external tool
+    for el in defs.assets or []:
+        if isinstance(el, CacheableAssetsDefinition):
+            raise DagsterInvalidDefinitionError(
+                "Can't use map_asset_specs on Definitions objects that contain "
+                "CacheableAssetsDefinitions."
+            )
+
+    new_keys_by_old_key: Dict[AssetKey, AssetKey] = {}
+    assets_def_new_specs: List[Tuple[AssetsDefinition, Sequence[AssetSpec]]] = []
+
+    assets_defs = defs.get_asset_graph().assets_defs
+    for el in assets_defs:
+        new_specs = []
+        for spec in el.specs:
+            new_spec = fn(spec)
+            new_specs.append(new_spec)
+            if new_spec.key != spec.key:
+                new_keys_by_old_key[spec.key] = new_spec.key
+        assets_def_new_specs.append((el, new_specs))
+
+    result_assets = []
+    result_asset_checks = []
+
+    for assets_def, new_specs in assets_def_new_specs:
+        new_asset_specs_replaced_deps = [
+            replace_asset_spec_dep_asset_keys(spec, new_keys_by_old_key) for spec in new_specs
+        ]
+        new_assets_def = assets_def.with_replaced_asset_specs(
+            specs=new_asset_specs_replaced_deps,
+            new_asset_keys_by_old_asset_key=new_keys_by_old_key,
+        )
+        if len(new_specs) > 0:
+            result_assets.append(new_assets_def)
+        else:
+            result_asset_checks.append(new_assets_def)
+
+    return copy(defs, assets=result_assets, asset_checks=result_asset_checks)
```

Comment:
I believe the asset check definitions are canonicalized to asset definitions in the guts of asset_graph. @smackesey to confirm

Reply:
I might just be confused by the variable names here.

Reply:
Actually now that I've looked at this this is bothering me quite a bit, as it speaks to the confusion about these "input" parameters. This actually has a bunch of inconsitent behavior. Splitting it back into `assets` and `asset_checks`  like this is a very imperfect heuristic, as a user can contruct an `AssetsDefinition` without an asset check and then it will end up back in the `asset_checks` parameter.

Furthermore, we're canonicalizing now anyways, so `assets` will not contain the same value.

I think the most consistent thing is to either try to preserve the original arguments or to fully canonicalize to `assets` for everything (including things passed to `asset_checks`). 

This speaks to the need to figure out this `record` thing for `Definitions`